internal/pkg/agent/cmd/container.go

```diff
@@ -988,3 +991,129 @@ func isContainer(detail component.PlatformDetail) component.PlatformDetail {
+	if os.IsNotExist(err) {
+		// config file does not exist so elastic-agent should enroll
+		return true, nil
+	}
+
+	// read the stored elastic-agent state to extract the previous fleet url and enrollment token
+	ctx := context.Background()
+	store, err := storage.NewEncryptedDiskStore(ctx, agentCfgFilePath)
+	if err != nil {
+		return false, fmt.Errorf("failed to instantiate encrypted disk store: %w", err)
+	}
+
+	reader, err := store.Load()
+	if err != nil {
+		return false, fmt.Errorf("failed to load from disk store: %w", err)
+	}
+
+	cfg, err := config.NewConfigFrom(reader)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedConfig, err := configuration.NewFromConfig(cfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedFleetHosts := storedConfig.Fleet.Client.GetHosts()
+	if len(storedFleetHosts) == 0 || !slices.Contains(storedFleetHosts, setupCfg.Fleet.URL) {
+		// fleet url changed thus elastic-agent should enroll
+		return true, nil
+	}
+
+	// here we don't return true if there is no previously saved enrollment token hash
+	// as an older version of elastic-agent won't have one.
+	if len(storedConfig.Fleet.EnrollmentTokenHash) > 0 {
+		// we have previously saved an enrollment token hash
+		enrollmentHashBytes, err := base64.StdEncoding.DecodeString(storedConfig.Fleet.EnrollmentTokenHash)
+		if err != nil {
+			return false, fmt.Errorf("failed to decode hash: %w", err)
+		}
+
+		err = crypto.ComparePBKDF2HashAndPassword(enrollmentHashBytes, []byte(setupCfg.Fleet.EnrollmentToken))
+		switch {
+		case errors.Is(err, crypto.ErrMismatchedHashAndPassword):
+			// the enrollment token hash does not match thus elastic-agent should enroll
+			return true, nil
+		case err != nil:
+			return false, fmt.Errorf("failed to compare hash: %w", err)
+		}
+	}
+
+	// Although fleet url and enrollment token are the same, we must also validate the api token.
+	// The agent might have been unenrolled from kibana which doesn't require for fleet url and
+	// enrollment token need to be changed
+	log, err := logger.New("fleet_client", false)
+	if err != nil {
+		return false, fmt.Errorf("failed to create logger: %w", err)
+	}
+	fc, err := fleetclient.NewAuthWithConfig(log, storedConfig.Fleet.AccessAPIKey, storedConfig.Fleet.Client)
+	if err != nil {
+		return false, fmt.Errorf("failed to create fleet client: %w", err)
+	}
+
+	// exploit the ACK request of fleet with **empty events**
+	// to validate the api token of the elastic-agent
+	ackRequest := &fleetapi.AckRequest{
+		Events: nil,
+	}
+	ackCMD := fleetapi.NewAckCmd(&agentInfo{storedConfig.Fleet.Info.ID}, fc)
+	_, err = ackCMD.Execute(ctx, ackRequest)
+	switch {
+	case errors.Is(err, fleetclient.ErrInvalidAPIKey):
+		// this elastic-agent has not valid api token thus it should enroll
+		return true, nil
+	case err != nil:
+		return false, fmt.Errorf("failed to validate api token: %w", err)
```

Comment:
A simple network error could cause this, should there be some retries in here?

Reply:
I added a really simple retry logic, let me know if that's align with what you had on mind 🙂

Reply:
@blakerouse any objections merging this PR?

Reply:
I am happy with the added retry loop. Would have been a little cleaner to place it in its own function, but that is okay. Feel free to merge.